### PR TITLE
test: Code Coverage: Add Edit Mileage #2 Part 1

### PR DESCRIPTION
**Pull request by jayfyle**
### <samp>🤖 Generated by Copilot at 53ac016</samp>

This pull request adds unit tests and mock data for the new mileage expense form feature. It creates a new file `add-edit-mileage-1.spec.ts` that tests the `AddEditMileageComponent` class, and modifies the existing file `org-category.data.ts` to include some sample mileage categories.

### <samp>🤖 Generated by Copilot at 53ac016</samp>

> _Oh we're the `AddEditMileageComponent` crew_
> _And we test our code with mock data too_
> _We heave and we ho on the count of three_
> _To make sure our mileage form works perfectly_

### <samp>🤖 Generated by Copilot at 53ac016</samp>

*  Add mock data of mileage categories for testing ([link](https://github.com/fylein/fyle-mobile-app/pull/2264/files?diff=unified&w=0#diff-8914bfb32bd70efe52e673bc9938855b2c84a494d9464f82c50fe2c8bfbdca45R1002-R1040))
*  Import mock data and other dependencies to the mileage component spec file ([link](https://github.com/fylein/fyle-mobile-app/pull/2264/files?diff=unified&w=0#diff-5ea8feba7a8355dcc3a878d59e6ac3c56d89e15472f7ddfe5d9c63c38a90c961R50-R60))
   - `goToTransaction` for navigating to the transaction page ([link](https://github.com/fylein/fyle-mobile-app/pull/2264/files?diff=unified&w=0#diff-5ea8feba7a8355dcc3a878d59e6ac3c56d89e15472f7ddfe5d9c63c38a90c961R300-R480))
   - `getProjectCategoryIds` for fetching the project and category ids from the form ([link](https://github.com/fylein/fyle-mobile-app/pull/2264/files?diff=unified&w=0#diff-5ea8feba7a8355dcc3a878d59e6ac3c56d89e15472f7ddfe5d9c63c38a90c961R300-R480))
   - `saveExpenseAndGotoNext` for saving the expense and moving to the next step ([link](https://github.com/fylein/fyle-mobile-app/pull/2264/files?diff=unified&w=0#diff-5ea8feba7a8355dcc3a878d59e6ac3c56d89e15472f7ddfe5d9c63c38a90c961R300-R480))
   - `getTimeSpentOnPage` for tracking the time spent on the form ([link](https://github.com/fylein/fyle-mobile-app/pull/2264/files?diff=unified&w=0#diff-5ea8feba7a8355dcc3a878d59e6ac3c56d89e15472f7ddfe5d9c63c38a90c961R300-R480))
   - `reloadCurrentRoute` for reloading the current route data ([link](https://github.com/fylein/fyle-mobile-app/pull/2264/files?diff=unified&w=0#diff-5ea8feba7a8355dcc3a878d59e6ac3c56d89e15472f7ddfe5d9c63c38a90c961R300-R480))
   - `continueWithPolicyViolations` for handling policy violations ([link](https://github.com/fylein/fyle-mobile-app/pull/2264/files?diff=unified&w=0#diff-5ea8feba7a8355dcc3a878d59e6ac3c56d89e15472f7ddfe5d9c63c38a90c961R300-R480))

## Clickup
https://app.clickup.com/t/85ztjmtf0

## Code Coverage
`13.63% Statements 117/858`
`4.05% Branches 28/690`
`8.53% Functions 25/293`
`14.18% Lines 116/818`

## UI Preview
Please add screenshots for UI changes

### CODE REVIEW

File: src/app/fyle/add-edit-mileage/add-edit-mileage-1.spec.ts [hunk 1]
```diff
@@ -286,5 +297,186 @@ export function TestCases1(getTestBed) {
         expect(trackingService.viewComment).toHaveBeenCalledOnceWith();
       }));
     });
+
+    describe('goToTransaction():', () => {
+      const txn_ids = ['txfCdl3TEZ7K'];
+      it('should navigate to add-edit mileage if category is mileage', () => {
+        const expense = { ...unflattenedTxnData, tx: { ...unflattenedTxnData.tx, org_category: 'MILEAGE' } };
```
Comment (OmkarJ13) [inline src/app/fyle/add-edit-mileage/add-edit-mileage-1.spec.ts:304]: Should we define this in mock data as well? Same for the per diem case

File: src/app/fyle/add-edit-mileage/add-edit-mileage-1.spec.ts [hunk 2]
```diff
@@ -286,5 +297,186 @@ export function TestCases1(getTestBed) {
         expect(trackingService.viewComment).toHaveBeenCalledOnceWith();
       }));
     });
+
+    describe('goToTransaction():', () => {
+      const txn_ids = ['txfCdl3TEZ7K'];
```
Comment (OmkarJ13) [inline src/app/fyle/add-edit-mileage/add-edit-mileage-1.spec.ts:302]: `txnIds` camelCase

File: src/app/fyle/add-edit-mileage/add-edit-mileage-1.spec.ts [hunk 3]
```diff
@@ -286,5 +297,186 @@ export function TestCases1(getTestBed) {
...
+      });
+    });
+
+    describe('getProjectCategoryIds():', () => {
+      it('should return MILEAGE category IDs', (done) => {
+        categoriesService.getAll.and.returnValue(of(mileageCategories));
+
+        component.getProjectCategoryIds().subscribe((res) => {
+          expect(res).toEqual(['141295', '141300']);
+          expect(categoriesService.getAll).toHaveBeenCalledOnceWith();
```
Comment (OmkarJ13) [inline src/app/fyle/add-edit-mileage/add-edit-mileage-1.spec.ts:358]: should we use toHaveBeenCalledTimes(1) here?

File: src/app/fyle/add-edit-mileage/add-edit-mileage-1.spec.ts [hunk 4]
```diff
@@ -286,5 +297,186 @@ export function TestCases1(getTestBed) {
...
+        expect(component.goToNext).toHaveBeenCalledTimes(1);
+      });
+    });
+
+    it('getTimeSpentOnPage(): should get time spent on page', () => {
+      component.expenseStartTime = 164577000;
+      fixture.detectChanges();
+
+      const result = component.getTimeSpentOnPage();
+      expect(result).toEqual((new Date().getTime() - component.expenseStartTime) / 1000);
```
Comment (OmkarJ13) [inline src/app/fyle/add-edit-mileage/add-edit-mileage-1.spec.ts:445]: Let's define this in another variable for readability
`new Date().getTime() - component.expenseStartTime) / 1000`

File: src/app/fyle/add-edit-mileage/add-edit-mileage-1.spec.ts [hunk 5]
```diff
@@ -286,5 +297,186 @@ export function TestCases1(getTestBed) {
...
+    });
+
+    it('reloadCurrentRoute(): should reload the current load', fakeAsync(() => {
+      component.reloadCurrentRoute();
+      tick(500);
```
Comment (OmkarJ13) [inline src/app/fyle/add-edit-mileage/add-edit-mileage-1.spec.ts:450]: Question: Why do we need this? Can we use `flushMicrotasks` instead?

Comment (OmkarJ13) [reply src/app/fyle/add-edit-mileage/add-edit-mileage-1.spec.ts:450]: I think it is recommended that for places where there is a specific timeout or intervals set, we should use `tick(x)` and for places where we've used async/await or promises, we should use `flushMicrotasks()`

From ChatGPT
```
tick:
Use tick when you need to simulate the passage of time, especially for operations scheduled using timers like setTimeout or setInterval.
tick is useful for testing time-dependent operations, such as functions that should be executed after a certain delay.
It's especially relevant for testing scenarios involving timers, animations, and other time-based behavior.

flushMicrotasks:
Use flushMicrotasks when you want to ensure that all microtasks (Promise callbacks) have been executed before moving on to the next step in your test.
It's useful for testing promises, async/await code, and ensuring that promise-based operations are completed.

```

File: src/app/core/mock-data/org-category.data.ts [hunk 6]
```diff
@@ -999,3 +999,42 @@ export const perDiemCategory: OrgCategory = {
   sub_category: 'Per Diem',
   updated_at: new Date('2022-09-13T17:16:56.232081+00:00'),
 };
+
+export const mileageCategories: OrgCategory[] = [
+  {
+    code: '93',
+    created_at: new Date('2021-05-18T11:40:38.576Z'),
+    displayName: 'Business',
```
Comment (Chethan-Fyle) [inline src/app/core/mock-data/org-category.data.ts:1007]: Should this be `displayName` or `display_name`?

File: src/app/fyle/add-edit-mileage/add-edit-mileage-1.spec.ts (hunk 3 above)
Comment (Chethan-Fyle) [reply src/app/fyle/add-edit-mileage/add-edit-mileage-1.spec.ts:358]: +1

**Review by Chethan-Fyle — APPROVED** [review]
minor comments.

File: src/app/core/mock-data/org-category.data.ts (hunk 6 above)
Comment (jayfyle) [reply src/app/core/mock-data/org-category.data.ts:1007]: It should be `displayName` only as defined in the model.

**Pull request merged**